---
title: "06a_gene_expression_analysis- dds7e: ~ Both_Resistance + Sample_Period + Species collaps Sample_period"
author: "Natalia Andrade Rodriguez"
date: "2023-03-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Gene expression analysis for all samples with at least 4M of mapped reads

###Library installing/loading
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
library(DESeq2)
library(ggplot2)
library(ggrepel)
library(dplyr)
library(tidyr)
library(reshape2)
library(tidyverse)
```
###Loading count data
```{r}

countData_4M<- readRDS("data/countData_4M.rds")

colnames_countData_4M<- as.data.frame(colnames(countData_4M)) %>% rename('colnames'='colnames(countData_4M)') %>% mutate(order = 1:144)
```
###Loading and sorting sample data
```{r}
samples_selected_4M<- read.csv("data/samples_selected_4M.csv")

sample_data_4M<- samples_selected_4M %>% 
  
              mutate(colnames = Sample_ID) %>% 
  
              full_join(colnames_countData_4M, by = 'colnames') %>% 
   
              mutate(SP_Resistance2 = paste(Sample_Period,Resistance2_April2021, sep="_" )) %>%
  
              mutate(SP_Resistance = paste(Sample_Period,Resistance_April2021, sep="_" )) %>%
  
              mutate(Both_Resistance = paste(Resistance2_April2021,Resistance_April2021, sep="_" )) %>% 
  
              mutate(SP_Both_Resistance = paste(Both_Resistance,Sample_Period, sep="_" )) %>% 
  
              arrange(order)
  
            
sample_data_4M$ID<- as.factor(sample_data_4M$ID)  
row.names(sample_data_4M)<- sample_data_4M$Sample_ID
sample_data_4M$Sample_Period<- as.factor(sample_data_4M$Sample_Period)  

keep <-countData_4M %>% mutate(keep = rowSums(countData_4M) >= 130)

countData_4M<- keep %>% filter(keep == TRUE) %>% select(-keep)


colnames(countData_4M) == rownames(sample_data_4M)

```
base on the model selection done in 6a_gene_expression_analysis.Rmd here we proceed to prepare the data to run that model

###Eliminating outliers
```{r}
countData_edit<- countData_4M %>% select(-'208_N56', -"272_N56", -"63_N56", -'172_LC053', -"342_LC053") 

keep <-countData_edit %>% mutate(keep = rowSums(countData_edit) >= 130)

countData_edit<- keep %>% filter(keep == TRUE) %>% select(-keep)

sample_data_edit<- sample_data_4M %>% filter(!grepl( '*_N56|*_LC053', colnames))
 
colnames(countData_edit) == rownames(sample_data_edit)
```

### dds7e --> ~ Both_Resistance + Sample_Period + Species, collapsed Sample_Period by ID and without outliers
```{r}
dds5e = DESeqDataSetFromMatrix(countData = countData_edit, 
                            
                                colData = sample_data_edit, 
 
                                design = ~ Both_Resistance + Sample_Period + Species)

dds7e<- collapseReplicates(dds5e, dds5e$ID, dds5e$Sample_Period)
colData(dds7e)
colnames(dds7e)
matchFirstLevel <- dds5e$ID == levels(dds5e$ID)[1]
```

```{r}
stopifnot(all(rowSums(counts(dds5e[,matchFirstLevel])) == counts(dds7e[,1])))
```

#Analysising DEG
```{r}
if (file.exists("cache/dds7e.rds")) {
  dds7e<- read_rds("cache/dds7e.rds")
} else {
  dds7e<- DESeq(dds7e)
  write_rds(dds7e,"cache/dds7e.rds")
}
```

```{r}
vsd7e = vst(dds7e, blind = FALSE)
```


#Ploting dds7e ~ Both_Resistance + Sample_Period + Species collaps Sample_Period as replicates of each ID WITHOUt outliers
```{r}
pcaData7e = plotPCA(vsd7e, intgroup=c("Species",'SP_Resistance',"Sample_ID"), 
returnData=TRUE)
percentVar7e = round(100 * attr(pcaData7e, "percentVar"))

#png("results/06a_modeltesting/PCA_dds7e_vst.png", width=7, height=7, units = "in", res = 300)
ggplot(pcaData7e, aes(PC1, PC2, colour = SP_Resistance, shape = Species)) + 
geom_point(size = 2) + theme_bw() + 
geom_text_repel(aes(label = Sample_ID), nudge_x = -1, nudge_y = 0.2, size = 3) +
ggtitle("Principal Component Analysis dds7e vst", subtitle = " ~ Both_Resistance + Sample_Period + Species collaps Sample_period NO OUTLIERS") +
xlab(paste0("PC1: ",percentVar7e[1],"% variance")) +
ylab(paste0("PC2: ",percentVar7e[2],"% variance"))
#dev.off()

#Dispersion plot
#png("results/06a_modeltesting/qc-dispersions_dds7e.png", 1000, 1000, pointsize=20)
plotDispEsts(dds7e, main="Dispersion plot  Both_Resistance + Sample_Period + Species collaps Sample_period NO OUTLIERS")
#dev.off()

#Cookâ€™s distance
#png("results/06a_modeltesting/cookdistance_dds7e.png", width=15, height=10, units = "in", res = 300)
boxplot_dds7e<-boxplot(log10(assays(dds7e)[["cooks"]]), range=0, las=0,  main="Cook's distance  Both_Resistance + Sample_Period + Species collaps Sample_period NO OUTLIERS")
#dev.off()
```

#transform data with rlogTransformation
```{r}
rld7e <- rlogTransformation(dds7e)
```

```{r}
pcaData7er = plotPCA(rld7e, intgroup=c("Species",'Both_Resistance',"Sample_ID"), 
returnData=TRUE)
percentVar7er = round(100 * attr(pcaData7er, "percentVar"))

png("results/06a_modeltesting/PCA_dds7e_rld.png", width=7, height=7, units = "in", res = 300)
ggplot(pcaData7er, aes(PC1, PC2, colour = Both_Resistance, shape = Species)) + 
geom_point(size = 2) + theme_bw() + 
geom_text_repel(aes(label = Sample_ID), nudge_x = -1, nudge_y = 0.2, size = 3) +
ggtitle("Principal Component Analysis dds7 rld", subtitle = " ~ Both_Resistance + Sample_Period + Species collaps Sample_period") +
xlab(paste0("PC1: ",percentVar7er[1],"% variance")) +
ylab(paste0("PC2: ",percentVar7er[2],"% variance"))
dev.off()

```

```{r}
png("hist_rld_dds7e.png", 1000, 1000, pointsize=20)
hist(assay(rld7e))
dev.off()
```

```{r}
resultsNames(dds7e)
res_dds7e<- results(dds7e)
summary(res_dds7e)
#attr(dds7e, "modelMatrix")
```
We will use contrast to compare our different groups of susceptibility 
###1: result we will compare "SCTLD_affected_Low" vs "SCTLD_unaffected_High"
```{r}
res_dds7e_1<- results(dds7e,contrast =
                                        
                            c("Both_Resistance",'SCTLD_affected_Low',"SCTLD_unaffected_High"))

resOrd_dds7e_1<- as.data.frame(res_dds7e_1[order(res_dds7e_1$padj),])
                
resOrd_dds7e_1<- resOrd_dds7e_1 %>%  
  
                 mutate(ofav_gene = rownames(resOrd_dds7e_1)) %>%
  
                 mutate(contrast = 1) %>% 
                 
                 mutate(contrast_gene = paste(contrast, ofav_gene, sep = "_")) %>% 
   
                 filter(padj<0.1)
```

###2: result we will compare "SCTLD_affected_Low" vs "SCTLD_affected_High"
```{r}
res_dds7e_2<- results(dds7e,contrast =
                                        
                            c("Both_Resistance",'SCTLD_affected_Low',"SCTLD_affected_High"))

resOrd_dds7e_2<- as.data.frame(res_dds7e_2[order(res_dds7e_2$padj),])
                
resOrd_dds7e_2<- resOrd_dds7e_2 %>%  
  
                 mutate(ofav_gene = rownames(resOrd_dds7e_2)) %>%
  
                 mutate(contrast = 2) %>% 
                 
                 mutate(contrast_gene = paste(contrast, ofav_gene, sep = "_")) %>% 
   
                 filter(padj<0.1)
```

###3: result we will compare "SCTLD_affected_Low" vs "SCTLD_affected_Some"
```{r}
res_dds7e_3<- results(dds7e,contrast =
                                        
                            c("Both_Resistance",'SCTLD_affected_Low',"SCTLD_affected_Some"))

resOrd_dds7e_3<- as.data.frame(res_dds7e_3[order(res_dds7e_3$padj),])
                
resOrd_dds7e_3<- resOrd_dds7e_3 %>%  
  
                 mutate(ofav_gene = rownames(resOrd_dds7e_3)) %>%
 
                 mutate(contrast = 3) %>% 
                 
                 mutate(contrast_gene = paste(contrast, ofav_gene, sep = "_")) %>% 
   
                 filter(padj<0.1)
```

###4: result we will compare "SCTLD_affected_Some" vs "SCTLD_unaffected_High"
```{r}
res_dds7e_4<- results(dds7e,contrast =
                                        
                            c("Both_Resistance",'SCTLD_affected_Some',"SCTLD_unaffected_High"))

resOrd_dds7e_4<- as.data.frame(res_dds7e_4[order(res_dds7e_4$padj),])
                
resOrd_dds7e_4<- resOrd_dds7e_4 %>%  
  
                 mutate(ofav_gene = rownames(resOrd_dds7e_4)) %>%
  
                 mutate(contrast = 4) %>% 
                 
                 mutate(contrast_gene = paste(contrast, ofav_gene, sep = "_")) %>% 
   
                 filter(padj<0.1)
```
###5: result we will compare "SCTLD_affected_Some" vs "SCTLD_affected_High"
```{r}
res_dds7e_5<- results(dds7e,contrast =
                                        
                            c("Both_Resistance","SCTLD_affected_Some" , "SCTLD_affected_High"))

resOrd_dds7e_5<- as.data.frame(res_dds7e_5[order(res_dds7e_5$padj),])
                
resOrd_dds7e_5<- resOrd_dds7e_5 %>%  
  
                 mutate(ofav_gene = rownames(resOrd_dds7e_5)) %>%
  
                 mutate(contrast = 5) %>% 
                 
                 mutate(contrast_gene = paste(contrast, ofav_gene, sep = "_")) %>% 
   
                 filter(padj<0.1)
```

###6: result we will compare "SCTLD_affected_High" vs "SCTLD_unaffected_High"
```{r}
res_dds7e_6<- results(dds7e,contrast =
                                        
                            c("Both_Resistance","SCTLD_affected_High" , "SCTLD_unaffected_High"))

resOrd_dds7e_6<- as.data.frame(res_dds7e_5[order(res_dds7e_6$padj),])
                
resOrd_dds7e_6<- resOrd_dds7e_6 %>%  
  
                 mutate(ofav_gene = rownames(resOrd_dds7e_6)) %>%
  
                 mutate(contrast = 6) %>% 
                 
                 mutate(contrast_gene = paste(contrast, ofav_gene, sep = "_")) %>% 
   
                 filter(padj<0.1)
```
###Merging all the list and adding annotations
```{r}
#Getting gene descriptions found with g:profiler

gene_anno<- read.csv("data/DEG_gprofiler.csv") %>% 

            mutate(contrast_gene = paste(contrast, ofav_gene, sep = "_")) %>% 
  
            select(-contrast, -ofav_gene)
#Binding results for all contrast

resOrd_dds7e_all<- rbind(resOrd_dds7e_1,resOrd_dds7e_2) %>% 
  
                   rbind(resOrd_dds7e_3,resOrd_dds7e_4) %>% 
  
                   rbind(resOrd_dds7e_5,resOrd_dds7e_6)

#Merging expression with annotations

DEG_all_dds7e<- resOrd_dds7e_all %>% full_join(gene_anno, by="contrast_gene")

write.csv(DEG_all_dds7e, "data/DEG_all_anno.csv")
```


```{r}

```

